overlord/snapstate/handlers.go

```diff
@@ -310,20 +323,120 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
 	return ts, err
 }
 
-func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq []string, userID int, tm timings.Measurer) error {
+// updates a prerequisite, if it's not providing a content interface that a plug expects it to
+func updatePrereqIfOutdated(t *state.Task, snapName string, contentAttrs []string, userID int, flags Flags) (*state.TaskSet, error) {
+	if len(contentAttrs) == 0 {
+		return nil, nil
+	}
+
+	st := t.State()
+
+	// check if the default provider has all expected content tags
+	if ok, err := hasAllContentAttrs(st, snapName, contentAttrs); err != nil {
+		return nil, err
+	} else if ok {
+		return nil, nil
```

Comment:
I think we are not hitting 'ok' case in the tests?

Reply:
No, because if hard to deterministically hit this case, but there is a `TestHasAllContentAttributes` in `snapstate_install_test.go` that should cover all of it. The issue is that the `ok` case will only be hit when the content is missing before the update tasks are scheduled but is then updated before the handler runs

Reply:
Wouldn't this case be hit when the provider snap has all the required content attributes and doesn't need updating, so after Install("consumer") we should see exactly one download happening for the consumer snap and not for the provider?

Reply:
The prerequisites->content map in the prereq task is only populated with content attributes that aren't already provided (I added some comments clarifying that in defaultProviderContentAttrs in an earlier commit). To trigger this case, the content provider can't already have this content at first so the prerequisites task gets a map with the content attributes and default providers to install, but before the prerequisites task of the consumer gets to run the producer is updated. This can happen because an update of the producer would not be prevent since the consumer's prerequisite task hasn't run and the producer install/update tasks haven't been scheduled. I couldn't think of a test that triggers this deterministically